### PR TITLE
V tagan/automation

### PR DESCRIPTION
**Pull request by TanujaKulal**
Device code auth flow using chrome driver is added to sample app.

### CODE REVIEW

File: automation/WinFormsAutomationApp/AuthenticationHelper.cs [hunk 1]
```diff
@@ -153,17 +145,66 @@ internal static class AuthenticationHelper
...
-                var list = TokenCache.DefaultShared.ReadItems().ToJson();
-                output.Add("AccessToken", list);
-                return output.ToJson();
-            });
+                //Get Device Code
+                DeviceCodeResult res1 = await ctx.AcquireDeviceCodeAsync(input["resource"], input["client_id"]);
 
-            return await myTask.ConfigureAwait(false);
+                //Open Chrome Driver in incognito mode
+                ChromeOptions options = new ChromeOptions();
```
Comment (mjcheetham) [inline automation/WinFormsAutomationApp/AuthenticationHelper.cs]: Everything from line 157 to line 185 here should be in the automation tests, not in the automation app.
`AcquireTokenUsingDeviceProfile` should only create the `AuthenticationContext`, call `ctx.AcquireDeviceCodeAsync` then return the result.

Comment (TanujaKulal) [inline automation/WinFormsAutomationApp/AuthenticationHelper.cs]: There was problem in context switching between chrome and sample app from automation for windriver. That is the reason we went and wrote this code in sample app.

Comment (mjcheetham) [inline automation/WinFormsAutomationApp/AuthenticationHelper.cs]: @TanujaKulal, that doesn't sound like a good reason to have all this code in the sample app itself?
If this code is working in the sample app then it should also work if it was in the test code.
What issues were you hitting when switching contexts? Perhaps we can help fix the issue in the test framework?

File: automation/WinFormsAutomationApp/packages.config [hunk 2]
```diff
@@ -0,0 +1,7 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="Selenium.Support" version="3.3.0" targetFramework="net452" />
```
Comment (mjcheetham) [inline automation/WinFormsAutomationApp/packages.config]: Please remove/undo adding these Selenium packages now that the browser automation bits have been moved into the TestAutomation repo.

File: automation/WinFormsAutomationApp/WinFormsAutomationApp.csproj [hunk 3]
```diff
@@ -109,6 +120,15 @@
     </ProjectReference>
   </ItemGroup>
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+  <Import Project="..\..\packages\Selenium.WebDriver.ChromeDriver.2.28.0\build\Selenium.WebDriver.ChromeDriver.targets" Condition="Exists('..\..\packages\Selenium.WebDriver.ChromeDriver.2.28.0\build\Selenium.WebDriver.ChromeDriver.targets')" />
```
Comment (mjcheetham) [inline automation/WinFormsAutomationApp/WinFormsAutomationApp.csproj]: Please remove these; no longer needed.

File: automation/WinFormsAutomationApp/WebDriverExtenstion.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+﻿using OpenQA.Selenium;
+using OpenQA.Selenium.Support.UI;
+using System;
+
+namespace WinFormsAutomationApp
+{
+    public static class WebDriverExtensions
+    {
+        public static IWebElement FindElement(this IWebDriver driver, By by, int timeoutInSeconds)
```
Comment (mjcheetham) [inline automation/WinFormsAutomationApp/WebDriverExtenstion.cs]: Shouldn't be needed anymore?

File: automation/WinFormsAutomationApp/DeleteCookies.cs [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        [DllImport(@"wininet",
+            SetLastError = true,
+            CharSet = CharSet.Auto,
+            EntryPoint = "DeleteUrlCacheEntryA",
+            CallingConvention = CallingConvention.StdCall)]
+        public static extern bool DeleteUrlCacheEntry(
+            IntPtr lpszUrlName);
+
+        [STAThread]
+        public static void CleanCoockies()
```
Comment (mjcheetham) [inline automation/WinFormsAutomationApp/DeleteCookies.cs]: Typo: `CleanCookies()`

File: automation/WinFormsAutomationApp/AuthenticationHelper.cs [hunk 6]
```diff
@@ -153,17 +164,58 @@ internal static class AuthenticationHelper
 
         public static async Task<string> AcquireTokenUsingDeviceProfile(Dictionary<string, string> input)
         {
-            Task<string> myTask = Task<string>.Factory.StartNew(() =>
+            Dictionary<string, object> res = new Dictionary<string, object>();
+            IWebDriver driver = null;
```
Comment (mjcheetham) [inline automation/WinFormsAutomationApp/AuthenticationHelper.cs]: Remove references to `IWebDriver`. No longer needed.

File: automation/WinFormsAutomationApp/AuthenticationHelper.cs [hunk 7]
```diff
@@ -153,17 +164,58 @@ internal static class AuthenticationHelper
...
+                res.Add("access_token", result.AccessToken);
+                res.Add("tenant_id", result.TenantId);               
+            }
+            catch (Exception exc)
+            {
+                res.Add("error", exc.Message);
+            }
+            finally
+            {
+               if(driver != null) driver.Quit();
```
Comment (mjcheetham) [inline automation/WinFormsAutomationApp/AuthenticationHelper.cs]: Not needed anymore.

File: automation/WinFormsAutomationApp/AuthenticationHelper.cs [hunk 8]
```diff
@@ -191,6 +243,16 @@ public static string ToJson(this object obj)
...
         #region Private Methods
+        // This function clears cookies from the browser control used by ADAL.
+        public static void ClearCookies()
+        {
+            const int INTERNET_OPTION_END_BROWSER_SESSION = 42;
+            InternetSetOption(IntPtr.Zero, INTERNET_OPTION_END_BROWSER_SESSION, IntPtr.Zero, 0);
+        }
+
+        [DllImport("wininet.dll", SetLastError = true)]
+        private static extern bool InternetSetOption(IntPtr hInternet, int dwOption, IntPtr lpBuffer, int lpdwBufferLength);
```
Comment (mjcheetham) [inline automation/WinFormsAutomationApp/AuthenticationHelper.cs]: Consider moving and consolidating all the P/Invoke code into a 'Native' directory/file or similar.

File: automation/WinFormsAutomationApp/AuthenticationHelper.cs [hunk 9]
```diff
@@ -294,7 +356,17 @@ private static void UpdateCache(KeyValuePair<TokenCacheKey, AuthenticationResult
             TokenCache.DefaultShared.StoreToCache(updated.Value, updated.Key.Authority, updated.Key.Resource, updated.Key.ClientId, updated.Key.TokenSubjectType, new CallState(new Guid()));
             NotifyAfterAccessCache(updated.Key.Resource, updated.Key.ClientId, updated.Value.Result.UserInfo.UniqueId, updated.Value.Result.UserInfo.DisplayableId);
         }
+
+        private static List<KeyValuePair<TokenCacheKey, AuthenticationResultEx>> QueryCache(string authority,
```
Comment (mjcheetham) [inline automation/WinFormsAutomationApp/AuthenticationHelper.cs]: Why a `List<KeyVaultPair<X,Y>>` and not a `Dictionary<X,Y>` ?

Comment (TanujaKulal) [inline automation/WinFormsAutomationApp/AuthenticationHelper.cs]: If it is KeyValuePair, updating cache becomes easier. no need to cast it back to keyvaluepair

**Pull request merged**